### PR TITLE
feat: project file uploader

### LINKED ISSUE #27926: New Project: File uploader site
## Project overview

File uploader site project for learners to practice Prisma ORM.

## Project outline

The following items define the scope of the project. The project outline is flexible; you can improve, expand, and omit items whilst writing lesson content (make sure the pull request description mentions that).

-  

## Acceptance criteria

If the requirements here are not met, the work effort is not complete.

- [ ] Lesson conforms to the outline above (unless outline has been modified).
- [ ] Lesson follows our [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md).

### Get the lesson template

Download the project template using the following command (replace `project_name` with the actual lesson name):

```bash
curl -o <lesson_name>.md https://raw.githubusercontent.com/TheOdinProject/curriculum/main/templates/project-template.md 
```

### PR DESCRIPTION
**Pull request by 01zulfi**
## Because

https://github.com/TheOdinProject/curriculum/issues/27926

## This PR

https://github.com/TheOdinProject/curriculum/issues/27926

## Issue

Closes #27926 

## Pull Request Requirements

-   [ ] I have thoroughly read and understand [The Odin Project curriculum contributing guide](https://github.com/TheOdinProject/curriculum/blob/main/CONTRIBUTING.md)
-   [ ] The title of this PR follows the `location of change: brief description of change` format, e.g. `Intro to HTML and CSS lesson: Fix link text`
-   [ ] The `Because` section summarizes the reason for this PR
-   [ ] The `This PR` section has a bullet point list describing the changes in this PR
-   [ ] If this PR addresses an open issue, it is linked in the `Issue` section
-   [ ] If any lesson files are included in this PR, they have been previewed with the [Markdown preview tool](https://www.theodinproject.com/lessons/preview) to ensure it is formatted correctly
-   [ ] If any lesson files are included in this PR, they follow the [Layout Style Guide](https://github.com/TheOdinProject/curriculum/blob/main/LAYOUT_STYLE_GUIDE.md)

### CODE REVIEW

File: nodeJS/orms/project_file_uploader.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+In this project, we'll be building a stripped down version of [Google Drive](https://www.google.com/drive/) (or any other personal storage service).
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. Set up a new project using Express and Prisma. Install all the necessary dependencies including Passport etc.
+1. Set up session based authentication using Passport.js. Use [Prisma session store library](https://github.com/kleydon/prisma-session-store#readme) to persist sessions in the database.
+1. Add a form where authenticated users can upload a file. Save the file in your filesystem for now. You'll need to integrate the [mutler middleware](https://github.com/expressjs/multer). We'll upload these files once we have all other features working.
```
Comment (jasonHYLam) [inline nodeJS/orms/project_file_uploader.md:11]: Small typo: mutler -> multer

**Review by jasonHYLam — COMMENTED** [review]
Small typo spotted on line 11, mutler -> multer

File: nodeJS/orms/project_file_uploader.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+1. Add a folders functionality. Users should be able to CRUD folders and upload files in it. Setup routes and necessary database interactions for this.
+1. Add a route to view a specific file details i.e. name, size, upload time. There should be a download button to allow users to download the file.
+1. Finally, add logic to upload files. You could store it in a database, but it's advised to use a cloud storage service for this usecase. You can use [Cloudinary](https://cloudinary.com/) or [Supabase storage](https://supabase.com/docs/guides/storage). When a file is uploaded, save the file URL in the database.
+
+#### Extra credit
+
+- Add a share folder functionality. When a user wants to share a folder (and all of its contents), they should have a form to specify the duration i.e. 1d, 10d etc. This should generate a link that can be shared with anyone (unauthenticated users). What the link would look like, `https://yourapp.com/share/c758c495-0705-44c6-8bab-6635fd12cf81`
```
Comment (Asartea) [inline nodeJS/orms/project_file_uploader.md:18]: This scans weird, maybe something like:
```suggestion
- Add a share folder functionality. When a user wants to share a folder (and all of its contents), they should have a form to specify the duration i.e. 1d, 10d etc. This should generate a link that can be shared with anyone (unauthenticated users). The link should be in the following format: `https://yourapp.com/share/c758c495-0705-44c6-8bab-6635fd12cf81`.
```

File: nodeJS/orms/project_file_uploader.md [hunk 3]
```diff
@@ -0,0 +1,20 @@
+### Introduction
+
+In this project, we'll be building a stripped down version of [Google Drive](https://www.google.com/drive/) (or any other personal storage service).
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. Set up a new project using Express and Prisma. Install all the necessary dependencies including Passport etc.
+1. Set up session based authentication using Passport.js. Use [Prisma session store library](https://github.com/kleydon/prisma-session-store#readme) to persist sessions in the database.
```
Comment (Asartea) [inline nodeJS/orms/project_file_uploader.md:10]:
```suggestion
1. Set up session based authentication using Passport.js. Use the [Prisma session store library](https://github.com/kleydon/prisma-session-store#readme) to persist sessions in the database.
```

File: nodeJS/orms/project_file_uploader.md [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+1. Add a form where authenticated users can upload a file. Save the file in your filesystem for now. You'll need to integrate the [multer middleware](https://github.com/expressjs/multer). We'll upload these files once we have all other features working.
+1. Add a folders functionality. Users should be able to CRUD folders and upload files in it. Setup routes and necessary database interactions for this.
+1. Add a route to view a specific file details i.e. name, size, upload time. There should be a download button to allow users to download the file.
```
Comment (Asartea) [inline nodeJS/orms/project_file_uploader.md:13]:
```suggestion
1. Add a route to view specific file details like name, size, and upload time. There should be a download button to allow users to download the file.
```

File: nodeJS/orms/project_file_uploader.md [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+In this project, we'll be building a stripped down version of [Google Drive](https://www.google.com/drive/) (or any other personal storage service).
+
+### Assignment
+
+<div class="lesson-content__panel" markdown="1">
+
+1. Set up a new project using Express and Prisma. Install all the necessary dependencies including Passport etc.
+1. Set up session based authentication using Passport.js. Use [Prisma session store library](https://github.com/kleydon/prisma-session-store#readme) to persist sessions in the database.
+1. Add a form where authenticated users can upload a file. Save the file in your filesystem for now. You'll need to integrate the [multer middleware](https://github.com/expressjs/multer). We'll upload these files once we have all other features working.
+1. Add a folders functionality. Users should be able to CRUD folders and upload files in it. Setup routes and necessary database interactions for this.
```
Comment (Asartea) [inline nodeJS/orms/project_file_uploader.md:12]:
```suggestion
1. Add folders. Users should be able to CRUD folders and upload files in them. Set up routes and necessary database interactions for this.
```

File: nodeJS/orms/project_file_uploader.md (hunk 2 above)
Comment (01zulfi) [reply nodeJS/orms/project_file_uploader.md]: I don't want to impose this format on the learners, I'm going with "For example, the link could be in the following format: ..."

**Review by Asartea — COMMENTED** [review]
LGTM. One additional idea for extra credit might be to also add a way to share permanently with some other user, but that is moreso a fun idea than anything I think the lesson really needs

**Comment by 01zulfi** [comment]
@Asartea Yep, sharing folders/files between registered users would be fun. I scraped that including the feature where multiple users could upload in a shared folder.  I'll add it if any other reviewer also wants it, but at this point in the curriculum I'd leave it to the learners on how feature rich they want their projects to be.

**Review by MaoShizhong — APPROVED** [review]
I like this. Don't think much needs to be added, and agreed that at this point, learners should be in charge of how feature rich they make things, we just provide the core requirements as part of the curriculum.

**Pull request merged**